### PR TITLE
Create import_region method for loading region objects

### PR DESCRIPTION
**Pull request by javerbukh**
### Description

This pull request creates an `import_region` method that loads a region object with `combination_mode`.
Some todo:
- [x] Implement loading region file using https://github.com/spacetelescope/jdaviz/blob/dfb4bfa9cf41a2129981c5c2f11ee5d4dca8677f/jdaviz/core/helpers.py#L651
- [x] `self.combination_mode` should use the `Select` (`SubsetSelect`?) class to be able to take advantage of the `choices` and `selected` attributes
- [x] Connect `self.combination_mode` to the subset mode selection in the toolbar
- [x] Add tests
- [x] Add documentation for new API

Fixes #<Issue Number>

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:681]: Why all the popping? Why not define these keywords explicitly instead of a blind `**kwargs`? Given that this is a public API, it is important for users to know all their options in the docstring as well (when you write the docstring that Kyle asked for).

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:691]:
```suggestion
                except Exception:  # nosec
```

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:720]: 1. Given that you `return` here, all the spatial region imports (if you still want them locally scoped) should happen after this.
2. You should `return bad_regions` here to be consistent with spatial loader behavior, which means your `_import_spectral_regions` should give you back `bad_regions` like the spatial one does.

```suggestion
            return self._import_spectral_regions(regions, mode)
```

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:829]: This behavior should be emulated by the spectral region loader too, to be consistent.

File: jdaviz/configs/default/plugins/subset_plugin/tests/test_subset_plugin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/subset_plugin/tests/test_subset_plugin.py:207]: Please write to `tmp_path`. Otherwise, this test will pollute the source dir when we run the tests locally.

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:126]: Is there a reason to keep the commented line?

```suggestion
```

File: jdaviz/configs/imviz/tests/test_regions.py
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_regions.py:192]: This changes the fundamental behavior of the test class. I wanted each method to have a new Imviz instance, but this change makes them all share the same instance. So hardcoded check like "Subset 1" is no longer reliable. Please revert this change.

File: jdaviz/core/helpers.py
Comment (pllim) [inline jdaviz/core/helpers.py:666]:
```suggestion
    @deprecated(since="4.0", alternative="viz.plugins['Subset Tools'].import_region")
```

File: jdaviz/core/helpers.py
Comment (pllim) [inline jdaviz/core/helpers.py:699]:
```suggestion
    @deprecated(since="4.0", alternative="viz.plugins['Subset Tools'].import_region")
```

File: notebooks/CubevizExample.ipynb
Comment (pllim) [inline notebooks/CubevizExample.ipynb:1]: Please revert your irrelevant changes here. Looks like you forgot to clear the notebook output before checking things in.

File: notebooks/SpecvizExample.ipynb
Comment (pllim) [inline notebooks/SpecvizExample.ipynb:1]: Please revert your irrelevant changes here. Looks like you forgot to clear the notebook output before checking things in.

**Review by pllim — CHANGES_REQUESTED** [review]
All the existing tests using `apply_roi` should be updated to use this new API too?

Is there any benefit to use locally scoped imports vs importing everything you need at the top of the module?

Other comments as follow.

**Comment by pllim** [comment]
Also:

1. Add a change log
2. Fix error in `jdaviz/configs/cubeviz/plugins/tests/test_tools.py`: `AssertionError: assert 3 == 4`

Thanks!

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:875]: this is only ever called when linking is changed, but needs to be two-way synced, so that changing the mode in the UI (the glue dropdown) or via internal glue API, is reflected in `self.combination_mode`

**Comment by javerbukh** [comment]
@pllim @kecnry I believe I have addressed all review comments except the ones about `_import_spectral_regions` behaving exactly like the spatial subset creation. I need to put this PR down until the 26th so please feel free to handle that additional work however you see fit.

**Comment by pllim** [comment]
> I need to put this PR down until the 26th

And this blocks 4.0?

**Comment by kecnry** [comment]
> And this blocks 4.0?

This is not listed as required for 4.0.  In fact, if we _do_ merge it, we may want to consider postponing exposing the API so that we continue to iterate on the public API during the follow-up efforts.

**Review by bmorris3 — COMMENTED** [review]
This looks good. Should we also change the example notebooks (ImvizExample, CubevizExample, ImvizDitheredExample, imviz_advanced_aper_phot, imviz_color_mpl, imviz_edit_subset_programmatic, cubeviz-cropped) which still have calls to `helper.load_regions`?

**Comment by javerbukh** [comment]
So add a `4.1 (unreleased)` to the `CHANGES.rst` file and a new 4.1 milestone?

**Comment by pllim** [comment]
The change log might have to wait, otherwise you will run into conflict. I can remilestone in the meantime.

**Comment by pllim** [comment]
I still find 144 instances of `apply_roi` that can be replaces with your new API. Do you have plans to address that in this PR? The point of this API is so user can stop using `apply_roi`, so I think our own code should be consistent internally.

```
$ grep apply_roi -rl jdaviz/
jdaviz/core/helpers.py
jdaviz/core/tools.py
jdaviz/core/tests/test_helpers.py
jdaviz/core/tests/test_data_menu.py
jdaviz/core/tests/test_template_mixin.py
jdaviz/tests/test_subsets.py
jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py
jdaviz/configs/cubeviz/plugins/tests/test_regions.py
jdaviz/configs/cubeviz/plugins/tests/test_parsers.py
jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_helper.py
jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py
jdaviz/configs/specviz/tests/test_helper.py
jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py
jdaviz/configs/default/plugins/data_quality/tests/test_data_quality.py
jdaviz/configs/default/plugins/model_fitting/tests/test_plugin.py
jdaviz/configs/default/plugins/export/tests/test_export.py
jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
jdaviz/configs/default/plugins/subset_plugin/tests/test_subset_plugin.py
```

**Comment by javerbukh** [comment]
This PR is not about replacing `apply_roi`, it is about creating an API to create subsets and import them from a file using the subset plugin. Please correct me if I'm wrong @kecnry @camipacifici .

**Comment by kecnry** [comment]
I personally think its ok if the base logic lives in an internal call.  The plugin provides a UI and API interface to the user, but that doesn't always make sense for all of our internal uses (and likely would add overhead).  That said, we probably should avoid duplicated logic whenever possible (I haven't looked at the diff to see if this is copying things from `apply_roi` or just adding a user-friendly wrapper around it.

If we want a consistent API for both internal calls and users, then we can mirror any changes to the calling signature of `apply_roi` or even call it something else, but I would be hesitant to make calls to a plugin from the app whenever possible.

**Comment by kecnry** [comment]
If we decide to replace `apply_roi` entirely, I think that could be a follow-up ticket, the scope of this has already extended quite a bit beyond the original scoped API.

**Comment by pllim** [comment]
(Shouldn't we be using the API that we want users to use in test code?)

**Comment by javerbukh** [comment]
I can't replace all calls to `apply_roi` but I can update the calls that are in tests.

**Comment by javerbukh** [comment]
@pllim @kecnry Tests have been updated and fixed, please review at your earliest convenience.

**Comment by javerbukh** [comment]
@kecnry I realized that many of the tests assume `viz.plugins['Subset Tools'].import_region` is public API so I reverted the commit. We can wait to merge this in that case.

**Comment by pllim** [comment]
Do we just want to branch v4.0.x now and then merge this into `main`?

**Comment by pllim** [comment]
My ticket too longer than expected. I will review Tuesday. Sorry for the delay.

**Comment by pllim** [comment]
Note to self: This shall not be merged until v4.0.x is branched out. Then Jesse is going to add change log to v4.1 section.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:838]: I remember @kecnry was concerned that `apply_roi` is no longer used within the public API to load spectral region, causing deviation between external and internal usage. Indeed, I do not see `_import_spectral_regions` use `apply_roi` at all because `_import_spectral_regions` is only called from image regions. So this concern still stands?

In fact, `create_new_after` is also not used?

File: jdaviz/core/helpers.py
Comment (pllim) [inline jdaviz/core/helpers.py:675]: When a change log is finally added, it needs to clarify that `max_num_regions=20` is no longer automatically set if loading image regions from file.

File: notebooks/concepts/imviz_edit_subset_programmatic.ipynb
Comment (pllim) [inline notebooks/concepts/imviz_edit_subset_programmatic.ipynb:95]: Where did the zoom level come from?

**Review by pllim — COMMENTED** [review]
1. Should `import_region` take ROI as input at all? ROI object is internal while the API is public. Or is this translation oversight? I think the ROI code should be converted to a proper circular region in the affected tests?
2. If the tests are meant to test the public API, should `subset_plugin = cubeviz_helper.plugins['Subset Tools']` be used instead of `subset_plugin = cubeviz_helper.plugins['Subset Tools']._obj` in the test code?
3. In the test code where `import_region(SpectralRegion(...))` is blindly using unit set in Unit Conversion plugin, I feel like the unit should be hardcoded instead. When this API is used, usually the user already have a unit in mind and won't refer to the Unit Conversion plugin at all. So reading the unit off Unit Conversion plugin does not reflect a realistic use case.

**Comment by camipacifici** [comment]
Hi @javerbukh, nice work!
I took this for a spin and have a couple of comments:
- For me `create_new_after=True` did nothing. Calling `import_region` would still overwrite the same subset with or without it. Maybe I am missing something.
- I tried the workflow where I create a list of regions and load all of them at once with
`regionlist = SpectralRegion([(1.5*u.um, 2.5*u.um),(3.5*u.um, 4.5*u.um)])`
`subtools.import_region(regionlist)`
Then I remembered I needed `subtools.combination_mode = 'new'` to be able to do this, which is fine, but I would argue that the default should be that after creating a subset, it creates a new one instead of overwriting the existing one.
What do you think?

**Comment by javerbukh** [comment]
@camipacifici For your second use case, doing `subtools.import_region(regionlist, create_new_after=True)` would do what you are expecting.

**Comment by camipacifici** [comment]
`create_new_after` does nothing on my side. I do not understand why. But I second Kyle's comment. It feels more natural to act on the current operation than on the next.

**Comment by javerbukh** [comment]
@camipacifici `create_new_after` has been renamed to `create_as_new` and now behaves as you would expect, including for spectral subsets. Please let me know if that works for you!

**Comment by camipacifici** [comment]
Thank you for the new updates!
I leave here something for discussion. My workflow is in specviz where I want to create three subsets
`region = SpectralRegion(1*u.um, 2*u.um) #alone`
`region2 = SpectralRegion(3*u.um, 4*u.um) #alone`
`regionlist = SpectralRegion([(1.5*u.um, 2.5*u.um),(3.5*u.um, 4.5*u.um)]) #2 wavelength ranges in same subset`
This does the job:
`subtools._obj.import_region(region, combination_mode='new')`
`subtools._obj.import_region(region2, combination_mode='new')`
`subtools._obj.import_region(regionlist, combination_mode=['new','or'])`
I wonder if `combination_mode='new'` should be the default for the first subset, to be able to achieve the same thing with this instead
`subtools._obj.import_region(region)`
`subtools._obj.import_region(region2)`
`subtools._obj.import_region(regionlist, combination_mode='or')`
What do people think?

Setting `subtools._obj.combination_mode = 'new'` at the beginning gives me a long traceback at the first `import_region`, but maybe I am doing something wrong there.

**Comment by javerbukh** [comment]
The first `import_region` call should create a new subset by default and all following calls will need the `combination_mode='new'` parameter. The reason for that is our `viewer.apply_roi` API behaves the same as the GUI where every call after the first just updates the selected subset. Creating a new subset with each call is what `helper.load_regions` does. This method can do both but the default glue and `apply_roi` implementation is also the default here.

**Comment by javerbukh** [comment]
Sorry, forgot to address this point! You need to do `subtools._obj.combination_mode.selected = 'new'` because `combination_mode` is a `SelectPluginComponent` object. You can also do `combination_mode.choices` to see options for what to select.

**Comment by kecnry** [comment]
ah ok, good to know its only that!.  (But note that once exposed to the user API, the `.selected` will no longer be required)

**Review by pllim — APPROVED** [review]
I think we're at the point where this has dragged on long enough that it should just go in so we can move on to the follow-up tickets. CI is green, so approving. Thanks for your patience and persistence!

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:67]: These aren't yet exposed in the user API

```suggestion
```

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:991]: has this been addressed?  When changing the combination mode from the UI, I do not see the internal traitlet being updated to match (although changing the `combination_mode.selected` does reflect in the UI via `_combination_selected_updated`).

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:995]: if this would ever need to be done in more than one place, we could have an inverted dictionary defined at the top (although then the dictionaries should probably be named something that makes it clear which is which direction).

```
SUBSET_TO_PRETTY = {v: k for k,v in SUBSET_MODES_PRETTY.items()}
```

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:839]: is there a reason these are not imported at the top-level?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:841]: why is SpectralRegion in this list?  Does it automatically handle iterating over itself?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:860]: is this a new TODO comment or copied from the previous region import code?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:881]: so if a list of regions is passed with a list of combination modes but an entry is invalid, the previous entries will be loaded and then an error will be raised here that it is not valid?  Can/should we validate the entire list in advance (possibly as follow-up if necessary)?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:721]: could/should we restore the original combination mode after this either completes or raises an exception?

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:839]: This is from the original code so probably a question for @pllim .

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:860]: Copied from previous, I'll remove it.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:881]: Validate the list of combination modes? Yes that should be a quick addition.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:721]: My thought would be the combination mode stays the last entry in the list but I am interested to hear what would make the most sense for a user.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:860]: does it work for cubes?  If not, let's make sure we have a ticket (but safe to ignore here if this was from moved code)

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:721]: ok, let's discuss and decide with the team.  Whatever the decision, its probably worth explaining here in the API docs so the users know what to expect.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:860]: Seems like this works, although the solution is hacky (per https://github.com/glue-viz/glue-astronomy/issues/75 ), maybe @pllim can clarify on the status. Subset creation works with cubes so I think this is safe to remove but I'll wait for Pey-Lian's answer.

Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py]: @rosteen reworked some spectrum WCS stuff recently so I'll defer to him. Or was it @dhomeier ? As long as the tests don't fail and results don't change, whatever works is fine by me.

**Comment by javerbukh** [comment]
@kecnry Waiting to hear from @rosteen on whether the line about checking a cube's wcs is still valid but otherwise I addressed the review comments. I tested changing the edit mode using the UI and API and it seemed to update correctly. Please let me know if you have a workflow I can use to test that behavior.

As for the comment on combination_mode reverting to default after `import_region` is called, I think we should defer that discussion to a follow-up. I created ticket for that in the subset plugin epic on Jira.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (rosteen) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py]: If it works it works 🤷 I don't recall changing the spectrum WCS stuff recently in a way that would affect this.

Comment (dhomeier) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py]: That issue has been closed by https://github.com/glue-viz/glue-astronomy/issues/75#issuecomment-2359390639, and the translator would only use a `PaddedSpectrumWCS` if the cube had a 1D WCS, which I assume is not the case here.

**Comment by kecnry** [comment]
After digging into this more, this might just be a limitation of how glue handles the mode dropdown vs the actual mode that will be used (when there is no active subset selected, there will be a new created subset regardless of the mode dropdown).   If/when we want to expose a UI element in the plugin, we'll definitely need to think about this a little more.  I'm not sure if/how we should deal with this for the exposed API though 🤔

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:62]:
```suggestion
COMBO_OPTIONS = list(SUBSET_MODES_PRETTY.keys())
```

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:139]:
```suggestion
        self.combination_mode = SelectPluginComponent(self,
                                                      items='combination_items',
                                                      selected='combination_selected',
                                                      manual_options=combo_options)
```

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:736]: I wonder if we could just do `data_label` - but we can revise this anytime before making public, so can ignore for now.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:764]: This is fine for now, but I wonder if there's any way we could predict in advance instead of try/except.  As it is now, checking the extension would probably work, but in the future we may have the same extension supported between spectral + spatial, so that isn't a longterm solution.

**Review by kecnry — APPROVED** [review]
LGTM now - any remaining items can be iterated as we build on top of this before we make the API exposed as public.  Thanks for working through all these iterations!

If we want this merged before @javerbukh gets back, I think we could just accept the few small code suggestions I listed and then merge.

**Comment by pllim** [comment]
> we could just accept the few small code suggestions I listed and then merge.

+1

**Pull request merged**
